src/mlpack/core/tree/vantage_point_tree/vantage_point_split.hpp

```diff
@@ -0,0 +1,197 @@
+/**
+ * @file vantage_point_split.hpp
+ * @author Mikhail Lozhnikov
+ *
+ * Definition of class VantagePointSplit, a class that splits a binary space
+ * partitioning into two parts using the distance to a certain vantage point.
+ */
+#ifndef MLPACK_CORE_TREE_BINARY_SPACE_TREE_VANTAGE_POINT_SPLIT_HPP
+#define MLPACK_CORE_TREE_BINARY_SPACE_TREE_VANTAGE_POINT_SPLIT_HPP
+
+#include <mlpack/core.hpp>
+
+namespace mlpack {
+namespace tree /** Trees and tree-building procedures. */ {
+
+template<typename BoundType, typename MatType = arma::mat>
+class VantagePointSplit
+{
+ public:
+  typedef typename MatType::elem_type ElemType;
+  /**
+   * Split the node according to the distance to a vantage point.
+   *
+   * @param bound The bound used for this node.
+   * @param data The dataset used by the binary space tree.
+   * @param begin Index of the starting point in the dataset that belongs to
+   *    this node.
+   * @param count Number of points in this node.
+   * @param splitCol The index at which the dataset is divided into two parts
+   *    after the rearrangement.
+   */
+  static bool SplitNode(const BoundType& bound,
+                        MatType& data,
+                        const size_t begin,
+                        const size_t count,
+                        size_t& splitCol);
+
+  /**
+   * Split the node according to the distance to a vantage point.
+   *
+   * @param bound The bound used for this node.
+   * @param data The dataset used by the binary space tree.
+   * @param begin Index of the starting point in the dataset that belongs to
+   *    this node.
+   * @param count Number of points in this node.
+   * @param splitCol The index at which the dataset is divided into two parts
+   *    after the rearrangement.
+   * @param oldFromNew Vector which will be filled with the old positions for
+   *    each new point.
+   */
+  static bool SplitNode(const BoundType& bound,
+                        MatType& data,
+                        const size_t begin,
+                        const size_t count,
+                        size_t& splitCol,
+                        std::vector<size_t>& oldFromNew);
+ private:
+  /**
+   * The maximum number of samples used for vantage point estimation and for
+   * estimation of the median.
+   */
+  static const size_t maxNumSamples = 100;
+
+  template<typename StructElemType>
+  struct SortStruct
+  {
+    size_t point;
+    ElemType dist;
+  };
+
+  template<typename StructElemType>
+  static bool StructComp(const SortStruct<StructElemType>& s1,
+      const SortStruct<StructElemType>& s2)
+  {
+    return (s1.dist < s2.dist);
+  };
```

Comment:
Should we use `std::pair<size_t, ElemType>` here too, like in the RectangleTree code?